File: sjob.py
#!/usr/bin/env -S python3 -u

import argparse
import glob
import os
import re
import shlex
import shutil
import subprocess
import sys
import time
from datetime import datetime


def parse_bool(value, default=False):
    if value is None:
        return default
    return str(value).strip().lower() in ("yes", "true", "1")


def parse_time_str(value, default_val=None):
    """Parses a time string (s, m, h, d) into seconds."""
    if not value:
        return default_val
    val = str(value).strip().lower()
    try:
        if val.endswith("d"):
            return int(val[:-1]) * 86400
        elif val.endswith("h"):
            return int(val[:-1]) * 3600
        elif val.endswith("m"):
            return int(val[:-1]) * 60
        elif val.endswith("s"):
            return int(val[:-1])
        else:
            return int(val)
    except ValueError:
        print(f"Error parsing time value '{value}'.")
        return default_val


def parse_start_hour(val):
    """Parses START_HOUR formats like '3', '2:45', or '1759' into seconds from midnight."""
    if not val:
        return None
    val = str(val).strip()
    hours, mins = 0, 0
    try:
        if ":" in val:
            h, m = val.split(":", 1)
            hours, mins = int(h), int(m)
        else:
            length = len(val)
            if length <= 2:
                hours, mins = int(val), 0
            elif length == 3:
                hours, mins = int(val[0]), int(val[1:])
            elif length == 4:
                hours, mins = int(val[:2]), int(val[2:])
            else:
                return None
    except ValueError:
        return None

    if 0 <= hours <= 23 and 0 <= mins <= 59:
        return (hours * 3600) + (mins * 60)
    return None


def parse_config(filepath):
    globals_cfg = {
        "DEBUG": False,
        "LOOP": 0,
        "TIMEFMT": "%Y-%m-%d %H:%M:%S",
        "DRY_RUN": False,
        "START_TIME_SEC": None,
    }
    jobs_cfg = {}

    # Regex to capture KEY=VALUE or KEY="VALUE"
    pattern = re.compile(r'^([A-Z_0-9]+)=[\'"]?(.*?)[\'"]?$')

    if not os.path.exists(filepath):
        print(f"Configuration file not found: {filepath}")
        sys.exit(1)

    with open(filepath, "r") as f:
        for line in f:
            line = line.strip()
            if not line or line.startswith("#"):
                continue

            match = pattern.match(line)
            if match:
                key, val = match.groups()

                # Check if it's a numbered job parameter
                m_suffix = re.search(r"(\d+)$", key)
                if m_suffix and any(
                    prefix in key for prefix in ("COPY_", "MOVE_", "REMOVE_", "CMD_")
                ):
                    idx = int(m_suffix.group(1))
                    base_key = key[: m_suffix.start()]

                    if idx not in jobs_cfg:
                        jobs_cfg[idx] = {}
                    jobs_cfg[idx][base_key] = val
                else:
                    # Global configuration
                    if key in ("START_HOUR", "START_TIME"):
                        globals_cfg["START_TIME_SEC"] = parse_start_hour(val)
                    elif key == "LOOP":
                        globals_cfg["LOOP"] = parse_time_str(val, 60)
                    elif key == "DELAY":
                        # Fallback to maintain backward compatibility
                        if globals_cfg["LOOP"] == 0:
                            globals_cfg["LOOP"] = parse_time_str(val, 60)
                    else:
                        globals_cfg[key] = val

    # Convert global types
    globals_cfg["DEBUG"] = parse_bool(globals_cfg.get("DEBUG"), False)
    globals_cfg["DRY_RUN"] = parse_bool(globals_cfg.get("DRY_RUN"), False)

    return globals_cfg, jobs_cfg


def log(msg, timefmt, strip_timestamp=False):
    if strip_timestamp:
        print(msg)
    else:
        now = datetime.now().strftime(timefmt)
        print(f"[{now}] {msg}")


def get_files(source_dir, expr, older_sec=None, newer_sec=None):
    """Fetches files matching glob, optionally filtering by age in seconds."""
    search_path = os.path.join(source_dir, expr)
    files = glob.glob(search_path)

    valid_files = []
    now = time.time()

    for f in files:
        if not os.path.isfile(f):
            continue

        if older_sec is None and newer_sec is None:
            valid_files.append(f)
            continue

        try:
            mtime = os.path.getmtime(f)
            age_in_seconds = now - mtime

            if older_sec is not None and age_in_seconds <= older_sec:
                continue

            if newer_sec is not None and age_in_seconds >= newer_sec:
                continue

            valid_files.append(f)
        except OSError:
            pass

    return valid_files


def run_hook(hook_name, cmd, debug, dry_run, timefmt, strip_timestamp):
    """Executes a job hook command safely if defined. Returns True on success."""
    if not cmd:
        return True

    if debug:
        log(f"Executing {hook_name} Hook: {cmd}", timefmt, strip_timestamp)

    if dry_run:
        return True

    result = subprocess.run(cmd, shell=True)
    return result.returncode == 0


def run_jobs(globals_cfg, jobs_cfg, strip_timestamp):
    debug = globals_cfg["DEBUG"]
    dry_run = globals_cfg["DRY_RUN"]
    timefmt = globals_cfg["TIMEFMT"]

    for idx in sorted(jobs_cfg.keys()):
        job = jobs_cfg[idx]

        # Determine job type dynamically
        if "COPY_JOB" in job:
            prefix = "COPY"
        elif "MOVE_JOB" in job:
            prefix = "MOVE"
        elif "REMOVE_JOB" in job:
            prefix = "REMOVE"
        elif "CMD_JOB" in job:
            prefix = "CMD"
        else:
            continue

        title = job.get(f"{prefix}_JOB", f"Job {idx}")
        source = job.get(f"{prefix}_SOURCE")
        expr = job.get(f"{prefix}_EXPR", "*")
        older = parse_time_str(job.get(f"{prefix}_OLDER"))
        newer = parse_time_str(job.get(f"{prefix}_NEWER"))
        pre_cmd = job.get(f"{prefix}_PRE")
        post_cmd = job.get(f"{prefix}_POST")

        # Specific mandatory constraints
        if prefix in ("COPY", "MOVE"):
            target = job.get(f"{prefix}_TARGET")
            if not source or not target:
                log(
                    f"WARNING: Job {idx} ({title}) missing mandatory SOURCE or TARGET. Skipping.",
                    timefmt,
                    strip_timestamp,
                )
                continue
        elif prefix == "REMOVE":
            if not source:
                log(
                    f"WARNING: Job {idx} ({title}) missing mandatory SOURCE. Skipping.",
                    timefmt,
                    strip_timestamp,
                )
                continue
        elif prefix == "CMD":
            command = job.get("CMD_COMMAND")
            if not source or not command:
                log(
                    f"WARNING: Job {idx} ({title}) missing mandatory SOURCE or COMMAND. Skipping.",
                    timefmt,
                    strip_timestamp,
                )
                continue

        files = get_files(source, expr, older, newer)

        if debug:
            age_flt = f" | Older: {older}s" if older else ""
            age_flt += f" | Newer: {newer}s" if newer else ""
            target_str = (
                f" | Target: {job.get(f'{prefix}_TARGET')}"
                if prefix in ("COPY", "MOVE")
                else ""
            )
            cmd_str = f" | Command: {job.get('CMD_COMMAND')}" if prefix == "CMD" else ""

            log(
                f"Starting {prefix} Job: '{title}' | Source: {source}{target_str}{cmd_str} | Expr: {expr}{age_flt}",
                timefmt,
                strip_timestamp,
            )
            log(
                f"Found {len(files)} files matching criteria.", timefmt, strip_timestamp
            )
            if files:
                log(
                    f"Files to process (showing up to 5): {', '.join(files[:5])}",
                    timefmt,
                    strip_timestamp,
                )
        else:
            log(f"Starting job: {title}", timefmt, strip_timestamp)

        # Execute PRE-execution hook command if specified.
        # Must return True (exit code 0) for job execution to continue.
        if not run_hook("PRE", pre_cmd, debug, dry_run, timefmt, strip_timestamp):
            log(
                f"WARNING: PRE hook failed for job '{title}'. Skipping job.",
                timefmt,
                strip_timestamp,
            )
            continue

        # Execute the core operations
        if prefix == "COPY":
            for f in files:
                if not dry_run:
                    os.makedirs(target, exist_ok=True)
                    shutil.copy2(f, target)

        elif prefix == "MOVE":
            for f in files:
                if not dry_run:
                    os.makedirs(target, exist_ok=True)
                    target_path = os.path.join(target, os.path.basename(f))
                    shutil.move(f, target_path)

        elif prefix == "REMOVE":
            for f in files:
                if not dry_run:
                    try:
                        os.remove(f)
                    except OSError as e:
                        log(f"Error removing file {f}: {e}", timefmt, strip_timestamp)

        elif prefix == "CMD":
            if not files:
                log(f"Job '{title}' done (no files).", timefmt, strip_timestamp)
                run_hook("POST", post_cmd, debug, dry_run, timefmt, strip_timestamp)
                continue

            replace_str = job.get("CMD_REPLACE", "{}")
            multiple = parse_bool(job.get("CMD_MULTIPLE"), False)

            if multiple:
                safe_files = " ".join([shlex.quote(f) for f in files])
                cmd_to_run = command.replace(replace_str, safe_files)
                if debug:
                    log(f"Executing: {cmd_to_run}", timefmt, strip_timestamp)
                if not dry_run:
                    subprocess.run(cmd_to_run, shell=True)
            else:
                for f in files:
                    cmd_to_run = command.replace(replace_str, shlex.quote(f))
                    if debug:
                        log(f"Executing: {cmd_to_run}", timefmt, strip_timestamp)
                    if not dry_run:
                        subprocess.run(cmd_to_run, shell=True)

        # Execute POST-execution hook command if specified
        run_hook("POST", post_cmd, debug, dry_run, timefmt, strip_timestamp)

        log(f"Job '{title}' done.", timefmt, strip_timestamp)


def main():
    parser = argparse.ArgumentParser(description="sjob File Processing Runner")
    parser.add_argument("config", help="Path to the configuration file")
    parser.add_argument(
        "-s",
        "--systemd",
        action="store_true",
        help="Disable showing internal script timestamps (useful for systemd/journalctl logs)",
    )

    args = parser.parse_args()

    globals_cfg, jobs_cfg = parse_config(args.config)
    timefmt = globals_cfg["TIMEFMT"]
    strip_ts = args.systemd
    loop_sec = globals_cfg["LOOP"]

    if globals_cfg["DEBUG"]:
        log("========= INITIALIZING JOB RUNNER =========", timefmt, strip_ts)
        log(f"LOOP: {loop_sec} seconds", timefmt, strip_ts)
        log("===========================================", timefmt, strip_ts)

    start_time_sec = globals_cfg.get("START_TIME_SEC")
    if start_time_sec is not None:
        now = datetime.now()
        current_sec = (now.hour * 3600) + (now.minute * 60) + now.second
        wait_sec = start_time_sec - current_sec

        if wait_sec <= 0:
            wait_sec += 86400  # 24 hours

        if globals_cfg["DEBUG"]:
            h = start_time_sec // 3600
            m = (start_time_sec % 3600) // 60
            log(
                f"START_HOUR/TIME set to {h:02d}:{m:02d}. Sleeping for {wait_sec} seconds before starting jobs.",
                timefmt,
                strip_ts,
            )
        else:
            log(
                f"Sleeping for {wait_sec} seconds until START_HOUR/TIME...",
                timefmt,
                strip_ts,
            )

        time.sleep(wait_sec)

    try:
        while True:
            if not jobs_cfg:
                log("No jobs found in configuration. Exiting.", timefmt, strip_ts)
                break

            loop_start_time = time.time()
            run_jobs(globals_cfg, jobs_cfg, strip_ts)

            if loop_sec > 0:
                elapsed_time = time.time() - loop_start_time
                wait_time = max(0, loop_sec - elapsed_time)

                if globals_cfg["DEBUG"]:
                    log(
                        f"All jobs finished. Time spent: {elapsed_time:.2f}s. Waiting for {wait_time:.2f}s before next loop...",
                        timefmt,
                        strip_ts,
                    )

                time.sleep(wait_time)
            else:
                log("LOOP is 0. Running once and exiting.", timefmt, strip_ts)
                break

    except KeyboardInterrupt:
        print("\nJob runner terminated by user.")
        sys.exit(0)


if __name__ == "__main__":
    main()

File: sjob.c
#include <ctype.h>
#include <errno.h>
#include <fcntl.h>
#include <glob.h>
#include <libgen.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>
#include <sys/time.h>
#include <sys/types.h>
#include <time.h>

// Handle getopt portability across Linux, macOS, and AIX
#if defined(__unix__) || defined(__AIX__) || defined(_AIX) || defined(__APPLE__)
#include <unistd.h>
#else
#include <getopt.h>
#endif

// For high-performance file copying where available
#if defined(__APPLE__)
#include <copyfile.h>
#elif defined(__linux__)
#include <sys/sendfile.h>
#endif

#define MAX_JOBS 1000
#define MAX_STR 1024
#define MAX_CMD 65536

typedef enum { JOB_NONE = 0, JOB_COPY, JOB_MOVE, JOB_REMOVE, JOB_CMD } JobType;

typedef struct {
  int id;
  JobType type;
  char title[MAX_STR];
  char source[MAX_STR];
  char target[MAX_STR];
  char expr[MAX_STR];
  int older;
  int newer;

  // CMD specific
  char command[MAX_STR];
  char replace[MAX_STR];
  bool multiple;
} Job;

typedef struct {
  bool debug;
  bool dry_run;
  int delay;
  char timefmt[MAX_STR];
  bool strip_ts;

  Job jobs[MAX_JOBS];
  int max_job_id;
} Config;

// Helpers
char *trim(char *str) {
  while (isspace((unsigned char)*str))
    str++;
  if (*str == 0)
    return str;
  char *end = str + strlen(str) - 1;
  while (end > str && isspace((unsigned char)*end))
    end--;
  end[1] = '\0';
  if ((str[0] == '"' && str[strlen(str) - 1] == '"') ||
      (str[0] == '\'' && str[strlen(str) - 1] == '\'')) {
    str[strlen(str) - 1] = '\0';
    str++;
  }
  return str;
}

bool parse_bool(const char *val, bool def) {
  if (!val || !*val)
    return def;
  if (strcasecmp(val, "yes") == 0 || strcasecmp(val, "true") == 0 ||
      strcmp(val, "1") == 0)
    return true;
  if (strcasecmp(val, "no") == 0 || strcasecmp(val, "false") == 0 ||
      strcmp(val, "0") == 0)
    return false;
  return def;
}

int parse_time_str(const char *val, int def) {
  if (!val || !*val)
    return def;
  int len = strlen(val);
  int multiplier = 1;
  char last = tolower((unsigned char)val[len - 1]);
  if (isalpha(last)) {
    if (last == 'd')
      multiplier = 86400;
    else if (last == 'h')
      multiplier = 3600;
    else if (last == 'm')
      multiplier = 60;
    else if (last == 's')
      multiplier = 1;
  }
  return atoi(val) * multiplier;
}

void log_msg(Config *cfg, const char *msg) {
  if (cfg->strip_ts) {
    printf("%s\n", msg);
  } else {
    time_t now = time(NULL);
    struct tm *t = localtime(&now);
    char t_str[MAX_STR];
    strftime(t_str, sizeof(t_str), cfg->timefmt, t);
    printf("[%s] %s\n", t_str, msg);
  }
  fflush(stdout);
}

void mkdir_p(const char *path) {
  char tmp[MAX_STR];
  snprintf(tmp, sizeof(tmp), "%s", path);
  size_t len = strlen(tmp);
  if (tmp[len - 1] == '/')
    tmp[len - 1] = 0;
  for (char *p = tmp + 1; *p; p++) {
    if (*p == '/') {
      *p = 0;
      mkdir(tmp, 0755);
      *p = '/';
    }
  }
  mkdir(tmp, 0755);
}

// Portable metadata cloning function
int clone_metadata(const char *src, const char *dst, struct stat *st) {
  // 1. Preserve Owner and Group
  if (chown(dst, st->st_uid, st->st_gid) != 0) {
    // Soft-fail: Ignore if running as non-root user
  }

  // 2. Preserve Permissions (chmod)
  if (chmod(dst, st->st_mode & 07777) != 0) {
    return -1;
  }

  // 3. Preserve Access and Modification Timestamps (utimes)
  struct timeval tv[2];
#if defined(__APPLE__)
  tv[0].tv_sec = st->st_atimespec.tv_sec;
  tv[0].tv_usec = st->st_atimespec.tv_nsec / 1000;
  tv[1].tv_sec = st->st_mtimespec.tv_sec;
  tv[1].tv_usec = st->st_mtimespec.tv_nsec / 1000;
#else
  tv[0].tv_sec = st->st_atime;
  tv[0].tv_usec = 0;
  tv[1].tv_sec = st->st_mtime;
  tv[1].tv_usec = 0;
#endif

  if (utimes(dst, tv) != 0) {
    return -1;
  }

  return 0;
}

// Enhanced file copy that preserves attributes
int copy_file_preserve(const char *src, const char *dst) {
  struct stat st;
  if (stat(src, &st) != 0)
    return -1;

#if defined(__APPLE__)
  if (copyfile(src, dst, NULL, COPYFILE_ALL) != 0)
    return -1;
  return 0;
#else
  int in_fd = open(src, O_RDONLY);
  if (in_fd < 0)
    return -1;

  int out_fd = open(dst, O_WRONLY | O_CREAT | O_TRUNC, 0666);
  if (out_fd < 0) {
    close(in_fd);
    return -1;
  }

#if defined(__linux__)
  off_t bytes_copied = 0;
  while (bytes_copied < st.st_size) {
    ssize_t res =
        sendfile(out_fd, in_fd, &bytes_copied, st.st_size - bytes_copied);
    if (res < 0) {
      if (errno == EINTR)
        continue;
      break;
    }
  }
#else
  char buf[16384];
  ssize_t n_read;
  while ((n_read = read(in_fd, buf, sizeof(buf))) > 0) {
    ssize_t n_written = 0;
    while (n_written < n_read) {
      ssize_t res = write(out_fd, buf + n_written, n_read - n_written);
      if (res < 0) {
        if (errno == EINTR)
          continue;
        break;
      }
      n_written += res;
    }
  }
#endif

  close(in_fd);
  close(out_fd);
  return clone_metadata(src, dst, &st);
#endif
}

int move_file_preserve(const char *src, const char *dst) {
  if (rename(src, dst) == 0)
    return 0;

  // Cross-device boundary shift fallback (EXDEV)
  if (errno == EXDEV) {
    if (copy_file_preserve(src, dst) == 0) {
      unlink(src);
      return 0;
    }
  }
  return -1;
}

void shell_quote(const char *src, char *dst, size_t dst_size) {
  size_t pos = 0;
  if (pos < dst_size - 1)
    dst[pos++] = '\'';
  while (*src && pos < dst_size - 5) {
    if (*src == '\'') {
      dst[pos++] = '\'';
      dst[pos++] = '\\';
      dst[pos++] = '\'';
      dst[pos++] = '\'';
    } else {
      dst[pos++] = *src;
    }
    src++;
  }
  if (pos < dst_size - 1)
    dst[pos++] = '\'';
  dst[pos] = '\0';
}

void str_replace(const char *orig, const char *rep, const char *with, char *dst,
                 size_t dst_size) {
  const char *tmp = orig;
  char *out = dst;
  size_t len_rep = strlen(rep);
  size_t len_with = strlen(with);

  while ((tmp = strstr(orig, rep)) != NULL) {
    size_t len_front = tmp - orig;
    if (out - dst + len_front + len_with >= dst_size - 1)
      break;
    strncpy(out, orig, len_front);
    out += len_front;
    strcpy(out, with);
    out += len_with;
    orig = tmp + len_rep;
  }
  strncpy(out, orig, dst_size - (out - dst) - 1);
  dst[dst_size - 1] = '\0';
}

void parse_config(const char *filepath, Config *cfg) {
  memset(cfg, 0, sizeof(Config));
  strcpy(cfg->timefmt, "%Y-%m-%d %H:%M:%S");

  for (int i = 0; i < MAX_JOBS; i++) {
    cfg->jobs[i].id = -1;
    strcpy(cfg->jobs[i].expr, "*");
    strcpy(cfg->jobs[i].replace, "{}");
  }

  FILE *f = fopen(filepath, "r");
  if (!f) {
    fprintf(stderr, "Configuration file not found: %s\n", filepath);
    exit(1);
  }

  char line[MAX_STR];
  while (fgets(line, sizeof(line), f)) {
    char *tline = trim(line);
    if (!*tline || *tline == '#')
      continue;

    char *eq = strchr(tline, '=');
    if (!eq)
      continue;

    *eq = '\0';
    char *key = trim(tline);
    char *val = trim(eq + 1);

    if (strcmp(key, "DEBUG") == 0)
      cfg->debug = parse_bool(val, false);
    else if (strcmp(key, "DRY_RUN") == 0)
      cfg->dry_run = parse_bool(val, false);
    else if (strcmp(key, "DELAY") == 0)
      cfg->delay = parse_time_str(val, 60);
    else if (strcmp(key, "TIMEFMT") == 0)
      strcpy(cfg->timefmt, val);
    else {
      char prefix[MAX_STR] = {0};
      int id = -1;
      int key_len = strlen(key);
      int num_idx = key_len - 1;
      while (num_idx >= 0 && isdigit(key[num_idx]))
        num_idx--;
      if (num_idx < key_len - 1) {
        id = atoi(&key[num_idx + 1]);
        strncpy(prefix, key, num_idx + 1);
        prefix[num_idx + 1] = '\0';
      }

      if (id >= 0 && id < MAX_JOBS) {
        Job *j = &cfg->jobs[id];
        j->id = id;
        if (id > cfg->max_job_id)
          cfg->max_job_id = id;

        if (strcmp(prefix, "COPY_JOB") == 0) {
          j->type = JOB_COPY;
          strcpy(j->title, val);
        } else if (strcmp(prefix, "COPY_SOURCE") == 0)
          strcpy(j->source, val);
        else if (strcmp(prefix, "COPY_TARGET") == 0)
          strcpy(j->target, val);
        else if (strcmp(prefix, "COPY_EXPR") == 0)
          strcpy(j->expr, val);
        else if (strcmp(prefix, "COPY_OLDER") == 0)
          j->older = parse_time_str(val, 0);
        else if (strcmp(prefix, "COPY_NEWER") == 0)
          j->newer = parse_time_str(val, 0);

        else if (strcmp(prefix, "MOVE_JOB") == 0) {
          j->type = JOB_MOVE;
          strcpy(j->title, val);
        } else if (strcmp(prefix, "MOVE_SOURCE") == 0)
          strcpy(j->source, val);
        else if (strcmp(prefix, "MOVE_TARGET") == 0)
          strcpy(j->target, val);
        else if (strcmp(prefix, "MOVE_EXPR") == 0)
          strcpy(j->expr, val);
        else if (strcmp(prefix, "MOVE_OLDER") == 0)
          j->older = parse_time_str(val, 0);
        else if (strcmp(prefix, "MOVE_NEWER") == 0)
          j->newer = parse_time_str(val, 0);

        else if (strcmp(prefix, "REMOVE_JOB") == 0) {
          j->type = JOB_REMOVE;
          strcpy(j->title, val);
        } else if (strcmp(prefix, "REMOVE_SOURCE") == 0)
          strcpy(j->source, val);
        else if (strcmp(prefix, "REMOVE_EXPR") == 0)
          strcpy(j->expr, val);
        else if (strcmp(prefix, "REMOVE_OLDER") == 0)
          j->older = parse_time_str(val, 0);
        else if (strcmp(prefix, "REMOVE_NEWER") == 0)
          j->newer = parse_time_str(val, 0);

        else if (strcmp(prefix, "CMD_JOB") == 0) {
          j->type = JOB_CMD;
          strcpy(j->title, val);
        } else if (strcmp(prefix, "CMD_SOURCE") == 0)
          strcpy(j->source, val);
        else if (strcmp(prefix, "CMD_COMMAND") == 0)
          strcpy(j->command, val);
        else if (strcmp(prefix, "CMD_EXPR") == 0)
          strcpy(j->expr, val);
        else if (strcmp(prefix, "CMD_REPLACE") == 0)
          strcpy(j->replace, val);
        else if (strcmp(prefix, "CMD_MULTIPLE") == 0)
          j->multiple = parse_bool(val, false);
        else if (strcmp(prefix, "CMD_OLDER") == 0)
          j->older = parse_time_str(val, 0);
        else if (strcmp(prefix, "CMD_NEWER") == 0)
          j->newer = parse_time_str(val, 0);
      }
    }
  }
  fclose(f);
}

void process_jobs(Config *cfg) {
  char msg[MAX_STR];
  time_t now = time(NULL);

  for (int i = 0; i <= cfg->max_job_id; i++) {
    Job *j = &cfg->jobs[i];
    if (j->id == -1 || j->type == JOB_NONE)
      continue;

    char search_path[MAX_STR];
    snprintf(search_path, sizeof(search_path), "%s/%s", j->source, j->expr);

    glob_t glob_res;
    int files_found = 0;
    char **valid_files = NULL;

    if (glob(search_path, 0, NULL, &glob_res) == 0) {
      valid_files = malloc(glob_res.gl_pathc * sizeof(char *));
      for (size_t g = 0; g < glob_res.gl_pathc; g++) {
        struct stat st;
        if (stat(glob_res.gl_pathv[g], &st) == 0 && S_ISREG(st.st_mode)) {
          int age = now - st.st_mtime;
          if (j->older > 0 && age <= j->older)
            continue;
          if (j->newer > 0 && age >= j->newer)
            continue;
          valid_files[files_found++] = strdup(glob_res.gl_pathv[g]);
        }
      }
    }

    if (cfg->debug) {
      snprintf(msg, sizeof(msg), "Starting Job: '%s' | Source: %s", j->title,
               j->source);
      log_msg(cfg, msg);
      snprintf(msg, sizeof(msg), "Found %d files matching criteria.",
               files_found);
      log_msg(cfg, msg);
    } else {
      snprintf(msg, sizeof(msg), "Starting job: %s", j->title);
      log_msg(cfg, msg);
    }

    if (j->type == JOB_COPY || j->type == JOB_MOVE) {
      if (!cfg->dry_run && files_found > 0)
        mkdir_p(j->target);
      for (int f = 0; f < files_found; f++) {
        char target_path[MAX_STR];
        const char *base = strrchr(valid_files[f], '/');
        base = base ? base + 1 : valid_files[f];

        snprintf(target_path, sizeof(target_path), "%s/%s", j->target, base);
        if (!cfg->dry_run) {
          if (j->type == JOB_COPY) {
            copy_file_preserve(valid_files[f], target_path);
          } else {
            move_file_preserve(valid_files[f], target_path);
          }
        }
      }
    } else if (j->type == JOB_REMOVE) {
      for (int f = 0; f < files_found; f++) {
        if (!cfg->dry_run)
          unlink(valid_files[f]);
      }
    } else if (j->type == JOB_CMD && files_found > 0) {
      if (j->multiple) {
        char *all_files = malloc(MAX_CMD);
        all_files[0] = '\0';
        for (int f = 0; f < files_found; f++) {
          char safe_path[MAX_STR];
          shell_quote(valid_files[f], safe_path, sizeof(safe_path));
          if (f > 0)
            strncat(all_files, " ", MAX_CMD - strlen(all_files) - 1);
          strncat(all_files, safe_path, MAX_CMD - strlen(all_files) - 1);
        }
        char cmd_to_run[MAX_CMD];
        str_replace(j->command, j->replace, all_files, cmd_to_run,
                    sizeof(cmd_to_run));

        if (cfg->debug) {
          char *cmd_msg = malloc(MAX_CMD + 50);
          if (cmd_msg) {
            snprintf(cmd_msg, MAX_CMD + 50, "Executing: %s", cmd_to_run);
            log_msg(cfg, cmd_msg);
            free(cmd_msg);
          }
        }
        if (!cfg->dry_run)
          system(cmd_to_run);
        free(all_files);
      } else {
        for (int f = 0; f < files_found; f++) {
          char safe_path[MAX_STR];
          shell_quote(valid_files[f], safe_path, sizeof(safe_path));
          char cmd_to_run[MAX_CMD];
          str_replace(j->command, j->replace, safe_path, cmd_to_run,
                      sizeof(cmd_to_run));

          if (cfg->debug) {
            char *cmd_msg = malloc(MAX_CMD + 50);
            if (cmd_msg) {
              snprintf(cmd_msg, MAX_CMD + 50, "Executing: %s", cmd_to_run);
              log_msg(cfg, cmd_msg);
              free(cmd_msg);
            }
          }
          if (!cfg->dry_run)
            system(cmd_to_run);
        }
      }
    }

    snprintf(msg, sizeof(msg), "Job '%s' done.", j->title);
    log_msg(cfg, msg);

    if (valid_files) {
      for (int f = 0; f < files_found; f++)
        free(valid_files[f]);
      free(valid_files);
    }
    globfree(&glob_res);
  }
}

int main(int argc, char *argv[]) {
  Config cfg;
  memset(&cfg, 0, sizeof(Config));
  cfg.strip_ts = false;

  int opt;
  while ((opt = getopt(argc, argv, "s")) != -1) {
    if (opt == 's')
      cfg.strip_ts = true;
  }

  if (optind >= argc) {
    fprintf(stderr, "Usage: %s [-s] <config_file>\n", argv[0]);
    return 1;
  }

  parse_config(argv[optind], &cfg);

  if (cfg.debug) {
    log_msg(&cfg, "=== INITIALIZING JOB RUNNER (C VERSION) ===");
    char msg[MAX_STR];
    snprintf(msg, sizeof(msg), "DELAY: %d seconds", cfg.delay);
    log_msg(&cfg, msg);
    log_msg(&cfg, "===========================================");
  }

  while (1) {
    if (cfg.max_job_id < 0) {
      log_msg(&cfg, "No jobs found in configuration. Exiting.");
      break;
    }

    time_t start_time = time(NULL);
    process_jobs(&cfg);

    if (cfg.delay > 0) {
      time_t elapsed = time(NULL) - start_time;
      int wait_time = cfg.delay - elapsed;
      if (wait_time < 0)
        wait_time = 0;

      if (cfg.debug) {
        char msg[MAX_STR];
        snprintf(msg, sizeof(msg),
                 "All jobs finished. Time spent: %lds. Waiting for %ds before "
                 "next loop...",
                 (long)elapsed, wait_time);
        log_msg(&cfg, msg);
      }
      sleep(wait_time);
    } else {
      log_msg(&cfg, "Delay is 0. Running once and exiting.");
      break;
    }
  }

  return 0;
}
